distributed/tests/test_client.py

```diff
@@ -4839,13 +4839,46 @@ async def kill(self, timeout=2, **kwargs):
 
 
 @pytest.mark.slow
+@pytest.mark.parametrize("raise_for_timeout", (True, False))
 @gen_cluster(client=True, Worker=SlowKillNanny)
-async def test_restart_workers_timeout(c, s, a, b):
-    with pytest.raises(TimeoutError) as excinfo:
-        await c.restart_workers(workers=[a.worker_address], timeout=0.001)
-    msg = str(excinfo.value).lower()
-    assert "workers failed to restart" in msg
-    assert a.worker_address in msg
+async def test_restart_workers_timeout(c, s, a, b, raise_for_timeout):
+    kwargs = dict(workers=[a.worker_address], timeout=0.001)
+
+    if raise_for_timeout:  # default is to raise
+        with pytest.raises(TimeoutError) as excinfo:
+            await c.restart_workers(**kwargs)
+        msg = str(excinfo.value).lower()
+        assert "workers failed to restart" in msg
+        assert a.worker_address in msg
+    else:
+        results = await c.restart_workers(raise_for_timeout=raise_for_timeout, **kwargs)
+        assert results == {a.worker_address: "timed out"}
+
+
+@pytest.mark.slow
+@gen_cluster(client=True, Worker=Nanny)
+async def test_restart_workers_exception(c, s, a, b):
+    async def fail_instantiate(*_args, **_kwargs):
+        raise ValueError()
+
+    a.instantiate = fail_instantiate
+
+    results = await c.restart_workers(workers=[a.worker_address])
+    msg: ErrorMessage = results[a.worker_address]
+    assert msg["status"] == "error"
+    assert msg["exception_text"] == "ValueError()"
```

Comment:
I believe it should be possible for us to actually raise this exception. The error_message also returns the proper exception type. 

What I'm thinking of is something like

```python
res = self.scheduler.restart_workers(...)
erred_workers = [msg for msg in res if msg['status'] != "OK"]
if erred_workers:
    # With py3.11 we might even want to use exception groups here
    first = erred_workers[0]
    raise first['exception']   
```

This is not a hard requirement so if this does not work as I expect it to, that's fine but I recommend checking it out quickly